### DISCUSSION TITLE
Let's Encrypt DNS-01 verification?

### DISCUSSION THREAD

**Discussion by gezusk** [Ylianst/MeshCentral#6075] [General]
With the recent changes by Let's Encrypt, I've had issues with updating my certs, because I geoblock all other countries. I recently converted my other sites to DNS verification using certbot. 

Is there a way to use DNS verification in MeshCentral? 

I've seen mention of setting up NGINX proxy for handling this. If I do that, what happens to the cert that MeshCentral currently has when it expires? Does it need to be removed before setting up the proxy?

**Comment by si458** [comment]
sadly meshcentral doesnt currently support dns verification, but the is an open issue about it being implemented
if you want to use a nginx proxy infront of meshcentral, the ssl that letsencrypt has will just expire on its own,
just make sure you set the reverse proxy settings in config.json correctly

**Comment by PathfinderNetworks** [comment]
For this very same reason I'd also like to see ZeroSSL support added to MeshCentral.  I mentioned it in another discussion a few weeks ago. 
https://github.com/Ylianst/MeshCentral/discussions/6057

**Reply by si458** [reply]
In theory all it requires is a server name change in the letsencrypt section but I will look into it this week for you!

**Reply by PathfinderNetworks** [reply]
It will also require that the ACME process respond to the EAB request from ZeroSSL.  I think that's the biggest difference (apart from the URI for the cert request server).

**Reply by si458** [reply]
From what I've quickly googled/developer docs, we already use the `acme-client` npm package which already supports all the EAB, so we just need options to set a server url and the end user to specify there EAB details which they can get from within the zerossl portal

**Reply by PathfinderNetworks** [reply]
That's exactly what I was hoping to hear.  It's pretty simple as long as the ACME client supports it.  The nice thing is ZeroSSL doesn't utilize request limits nor does it use the random/spread all over the world verification servers like LetsEncrypt.  The lack of request limits makes it MUCH more user friendly when it comes to troubleshooting. And you can log in to your ZeroSSL account at their site and see all of your existing certificates and their status.  Now that I've been using it I much prefer it over LetsEncrypt.   
If I knew anything about code at all I'd help out.  But the most I can handle is making config changes to existing json files.  One of these days maybe I'll try to learn how to code (like once I retire from my IT services business).

**Reply by si458** [reply]
Don't worry I'll look into it as it does look simple enough! Was just thinking adding extra values in the letsencrypt section of config.json, like ead details and serverurl and if they specified then use them, so it's universal for other acme servers too!

**Comment by gezusk** [comment]
I got certbot installed, and used it to certonly a certificate, and then used a deploy-hook to copy the files to the meshcentral-data folder. The only think I noticed is that I have to restart the meshcentral service to see the cert change on the website. Is that necessary? Would it eventually pick it up, or do I need to add to the deploy-hook to restart it?

**Reply by mtran-88** [reply]
Hi, would you be able to share how you configure MeshCentral to use ZeroSSL, and can it be used if I've already had a domain name?  What is certbot, and deploy-hook?  Also, how would I integrate with ACME?  Thx!

**Reply by brianbauer42** [reply]
@gezusk I'm also interested in hearing about your configuration. I am trying to deploy MeshCentral in a home k8s cluster where I already have certbot running for traefik, but am still pretty new to the concepts involved. I'll look into deploy hooks, but it might be helpful to see how it is configured. 

I'm double NAT'd and behind CGNAT, so DNS-01 letsencrypt is the only good option for me at the moment and from this thread it sounds like it isn't supported.

**Reply by gezusk** [reply]
@mtran-88 I used certbot with Let's Encrypt and Cloudflare
@brianbauer42 I used certbot with the cloudflare plugin. I did certonly, and it stores the cert in a folder. I then created a .sh file that has file copy commands, which I called with certbot's deploy-hook feature. So, when the certbot process runs, and generates new files, the deploy-hook features runs the .sh file, and copies the files to replace the ones in the MeshCentral folder.

What I'm not sure about is if restarting the MeshCentral process is necessary. During testing, MeshCentral didn't pick up the new cert until I restarted the process. So, I may have to add a line to the .sh file to do that. Or, maybe I'll just schedule a weekly restart.

**Reply by PetieM** [reply]
Hey @gezusk, did you ever determine if the restart was necessary? I'm currently just fine using the built-in Let's Encrypt functionality but all of my other services use DNS verification via Cloudflare so if I could get my Mesh Central server set up the same way reliably, that would be great. Thanks!

**Comment by gezusk** [comment]
Yes, the Mesh Central service needs to be restarted. Funny enough, new
versions have been coming out often enough that it hasn't been an issue
yet. But restarting the service makes it pick up the new certificate.

**Reply by PetieM** [reply]
Sweet, thanks! So just to confirm, you renew via certbot, use the deploy-hook to run a script that copies the certificate files from the letsencrypt folder to meshcentral-data, and then it restarts the service?

**Comment by gezusk** [comment]
Well, it copies the file, but it doesn't restart the service. I haven't
tried fixing that though, since it's hasn't been an issue yet 😁

**Reply by PetieM** [reply]
Fair enough! I appreciate the reply a full year after the initial comments were made. Have a good one!
